Increase web worker initial js heap size

Andrew-Fryer opened the pull request:
#### e24440a867d4507df732f66e1be534139414a9da
<pre>
Increase web worker initial js heap size
<a href="https://bugs.webkit.org/show_bug.cgi?id=290687">https://bugs.webkit.org/show_bug.cgi?id=290687</a>
<a href="https://rdar.apple.com/148155607">rdar://148155607</a>

Reviewed by Yusuke Suzuki.

Increase web worker js heap initial size from 1 mb to 4mb.
Introduce a new HeapType (medium) used by web workers
so that daemon clients aren&apos;t affected.

* Source/JavaScriptCore/heap/Heap.cpp:
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/WebCore/workers/WorkerGlobalScope.cpp:
(WebCore::WorkerGlobalScope::WorkerGlobalScope):

Canonical link: <a href="https://commits.webkit.org/292912@main">https://commits.webkit.org/292912@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b7484d6e12f4e89b79cfed326bdfd82e1a4327d8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97425 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17050 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7266 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102512 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47953 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99470 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17343 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25501 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74238 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/31420 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100428 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13131 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88112 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54582 "Found 1 new API test failure: /WPE/TestSSL:/webkit/WebKitWebView/ephemeral-tls-errors (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12897 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5997 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47396 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/90101 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82929 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6078 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104532 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/96047 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24504 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/17872 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83285 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/24876 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84241 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82706 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27247 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4935 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/18053 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15741 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24466 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29634 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/119673 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24288 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/33590 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27602 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25862 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/heap/Heap.cpp:
```diff
@@ -128,6 +128,8 @@ size_t minHeapSize(HeapType heapType, size_t ramSize)
         return static_cast<size_t>(result);
+    } else if (heapType == HeapType::Medium) {
+        return static_cast<size_t>(Options::mediumHeapSize());
     }
```
WebKit style is no braces for one line and no `else` if the previous `if` always returns.

Constellation replied:
I think probably we should use switch to list up all possible heapType here.
```
switch (heapType) {
case HeapType::Large:
   ...
case HeapType::Medium:
   ...
case HeapType::Small:
   ...
default:
   RELEASE_ASSERT_NOT_REACHED();
   break;
}
```

Constellation reviewed (approved):
r=me with comments.

Andrew-Fryer commented on Source/JavaScriptCore/heap/Heap.cpp:
```diff
@@ -123,13 +123,19 @@ double maxPauseMS(double thisPauseMS)
             static_cast<double>(Options::largeHeapSize()),
-            ramSize * Options::smallHeapRAMFraction());
-        return static_cast<size_t>(result);
+            ramSize * Options::smallHeapRAMFraction()));
```
The style checker doesn't like this, so I had to force it.
I think it is just confused (maybe due to `Options::...` starting with a capital?).
Please let me know if I should do this differently to appease the style checker.
```
Andrews-MacBook-Pro:JavaScriptCore andrewfryer$ git webkit pr
ERROR: Source/JavaScriptCore/heap/Heap.cpp:130:  Declaration has space between type name and * in ramSize * Options  [whitespace/declaration] [3]
Total errors found: 1 in 4 files
style-checker failed!
Retry will amend the commit with your changes. Continue uploading pull request? (Yes/Retry/[No]): Yes
```

kmiller68 replied:
I think it's because the regex for the style checker thinks you're declaring a variable `Options` of type `ramSize*` but you put a space between the type name and the star. Anyway it's confused so just ignore it.

Andrew-Fryer committed:
Increase web worker initial js heap size
https://bugs.webkit.org/show_bug.cgi?id=290687
rdar://148155607

Reviewed by Yusuke Suzuki.

Increase web worker js heap initial size from 1 mb to 4mb.
Introduce a new HeapType (medium) used by web workers
so that daemon clients aren't affected.

* Source/JavaScriptCore/heap/Heap.cpp:
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/WebCore/workers/WorkerGlobalScope.cpp:
(WebCore::WorkerGlobalScope::WorkerGlobalScope):

Canonical link: https://commits.webkit.org/292912@main
